File: ios/Views/ProgressiveBlurView.swift
import ExpoModulesCore
import UIKit

class ProgressiveBlurView: ExpoView {
  private static let fadeSamples = 5

  private static let blurScale: CGFloat = 1.25
  private static let blurRampShare: CGFloat = 0.6
  private static let coverageFadeDistance: CGFloat = 24
  private static let scrollViewCheckInterval: TimeInterval = 0.5
  private static let defaultWashOpacity: CGFloat = 0.85

  private let blurContainer = UIView()
  private let maskedBlur = UIView()
  private let effectView = BlurEffectView()
  private let blurMask = GradientView()
  private let tintWash = GradientView()
  private let fallbackGradient = GradientView()
  private let scrollMonitor = ScrollBlurMonitor()

  private var liveBlurAmount: CGFloat = 1
  private var scrollViewCheck: Timer?

  var intensity: Double = 25 {
    didSet {
      effectView.intensity = Double(blurStrength)
      updateGradients()
      updateLayerVisibility()
    }
  }

  var tint: TintStyle = .systemUltraThinMaterial {
    didSet { effectView.tint = tint }
  }

  var blurTintColor: UIColor? {
    didSet { updateGradients() }
  }

  var edge: ProgressiveBlurEdge = .top {
    didSet { setNeedsLayout() }
  }

  var startOffset: Double = 0 {
    didSet { setNeedsLayout() }
  }

  var scrollFallback: Bool = true {
    didSet {
      scrollMonitor.tracksSpeed = scrollFallback
      updateLayerVisibility()
    }
  }

  var fallbackColor: UIColor? {
    didSet { updateGradients() }
  }

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)

    blurContainer.isUserInteractionEnabled = false

    blurContainer.layer.allowsGroupOpacity = false
    effectView.tint = tint
    effectView.intensity = Double(blurStrength)
    maskedBlur.addSubview(effectView)
    maskedBlur.layer.mask = blurMask.layer
    blurContainer.addSubview(maskedBlur)
    tintWash.isHidden = true
    blurContainer.addSubview(tintWash)
    addSubview(blurContainer)

    fallbackGradient.isHidden = true
    addSubview(fallbackGradient)

    scrollMonitor.tracksSpeed = scrollFallback
    scrollMonitor.onScroll = { [weak self] in
      self?.updateLayerVisibility()
    }
    scrollMonitor.onContentChange = { [weak self] in
      self?.updateLayerVisibility()
    }
    scrollMonitor.onBlurAmountChange = { [weak self] amount in
      self?.liveBlurAmount = amount
      self?.updateLayerVisibility()
    }
  }

  deinit {
    scrollViewCheck?.invalidate()
  }

  override func mountChildComponentView(_ childComponentView: UIView, index: Int) {
    insertSubview(childComponentView, at: index + 2)
  }

  override func unmountChildComponentView(_ childComponentView: UIView, index: Int) {
    childComponentView.removeFromSuperview()
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    blurContainer.frame = bounds
    maskedBlur.frame = bounds
    effectView.frame = bounds
    blurMask.frame = bounds
    tintWash.frame = bounds
    fallbackGradient.frame = bounds
    updateGradients()

    if scrollMonitor.scrollView == nil {
      attachScrollMonitor()
    }
    updateLayerVisibility()
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    attachScrollMonitor()

    scrollViewCheck?.invalidate()
    scrollViewCheck = nil
    guard window != nil else { return }

    let timer = Timer(timeInterval: Self.scrollViewCheckInterval, repeats: true) { [weak self] _ in
      self?.reattachIfScrollViewLeft()
    }
    RunLoop.main.add(timer, forMode: .default)
    scrollViewCheck = timer
  }

  override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
    super.traitCollectionDidChange(previousTraitCollection)
    updateGradients()
  }

  private var fadeLength: CGFloat {
    1 - CGFloat(max(0, min(1, startOffset)))
  }

  private var strength: CGFloat {
    CGFloat(max(0, min(100, intensity)) / 100)
  }

  private var blurStrength: CGFloat {
    min(1, strength * Self.blurScale)
  }

  private func updateGradients() {
    let blurRising = Self.fade(from: 0, to: fadeLength * Self.blurRampShare, rising: true) + [(1, 1)]
    blurMask.update(edge: edge, color: .black, stops: blurRising)

    let rising = Self.fade(from: 0, to: fadeLength, rising: true) + [(1, 1)]

    let background = Self.backgroundColor(behind: self)

    let wash = (blurTintColor ?? background.withAlphaComponent(Self.defaultWashOpacity))
      .resolvedColor(with: traitCollection)
    tintWash.isHidden = wash.cgColor.alpha == 0
    tintWash.update(edge: edge, color: wash, stops: rising)

    let fallback = (fallbackColor ?? background).resolvedColor(with: traitCollection)
    let stops: [GradientStop] = rising.map { ($0.location, $0.alpha * strength) }
    fallbackGradient.update(edge: edge, color: fallback, stops: stops)
  }

  private static func fade(from start: CGFloat, to end: CGFloat, rising: Bool) -> [GradientStop] {
    (0...fadeSamples).map { sample in
      let t = CGFloat(sample) / CGFloat(fadeSamples)
      let eased = smoothstep(0, 1, t)
      return (start + (end - start) * t, rising ? eased : 1 - eased)
    }
  }

  private static func backgroundColor(behind view: UIView) -> UIColor {
    var ancestor = view.superview
    while let current = ancestor {
      if let color = current.backgroundColor, color.cgColor.alpha >= 1 {
        return color
      }
      ancestor = current.superview
    }
    #if os(tvOS)
      return view.window?.backgroundColor ?? .black
    #else
      return view.window?.backgroundColor ?? .systemBackground
    #endif
  }

  private func updateLayerVisibility() {
    let coverage = contentCoverage()
    let blurAmount = scrollFallback ? liveBlurAmount : 1

    let blurAlpha = blurAmount * coverage
    if blurContainer.alpha != blurAlpha {
      blurContainer.alpha = blurAlpha
    }
    let blurHidden = blurAlpha == 0 || strength == 0
    if blurContainer.isHidden != blurHidden {
      blurContainer.isHidden = blurHidden
    }

    let fallbackAlpha = (1 - blurAmount) * coverage
    if fallbackGradient.alpha != fallbackAlpha {
      fallbackGradient.alpha = fallbackAlpha
    }
    let fallbackHidden = fallbackAlpha == 0
    if fallbackGradient.isHidden != fallbackHidden {
      fallbackGradient.isHidden = fallbackHidden
    }
  }

  private func contentCoverage() -> CGFloat {
    guard let scrollView = scrollMonitor.scrollView else { return 1 }
    let inset = scrollView.adjustedContentInset
    let offset = scrollView.contentOffset
    let size = scrollView.contentSize
    let visible = scrollView.bounds.size

    let hiddenContent: CGFloat
    switch edge {
    case .top:
      hiddenContent = offset.y + inset.top
    case .bottom:
      hiddenContent = size.height + inset.bottom - (offset.y + visible.height)
    case .left:
      hiddenContent = offset.x + inset.left
    case .right:
      hiddenContent = size.width + inset.right - (offset.x + visible.width)
    }
    return smoothstep(0, Self.coverageFadeDistance, hiddenContent)
  }

  private func attachScrollMonitor() {
    guard window != nil else {
      scrollMonitor.stop()
      return
    }
    scrollMonitor.observe(findScrollViewBehind())
    updateGradients()
    updateLayerVisibility()
  }

  private func reattachIfScrollViewLeft() {
    guard let scrollView = scrollMonitor.scrollView, scrollView.window == nil else { return }
    scrollMonitor.stop()
    attachScrollMonitor()
  }

  private func findScrollViewBehind() -> UIScrollView? {
    let frameInWindow = convert(bounds, to: nil)
    var child: UIView = self

    while let container = child.superview {
      let index = container.subviews.firstIndex(of: child) ?? 0
      for sibling in container.subviews[..<index].reversed() {
        if let scrollView = Self.scrollView(in: sibling, overlapping: frameInWindow) {
          return scrollView
        }
      }
      child = container
    }
    return nil
  }

  private static func scrollView(in view: UIView, overlapping frame: CGRect) -> UIScrollView? {
    guard !view.isHidden, view.convert(view.bounds, to: nil).intersects(frame) else { return nil }
    if let scrollView = view as? UIScrollView {
      return scrollView
    }
    for subview in view.subviews.reversed() {
      if let scrollView = scrollView(in: subview, overlapping: frame) {
        return scrollView
      }
    }
    return nil
  }
}

private typealias GradientStop = (location: CGFloat, alpha: CGFloat)

private final class GradientView: UIView {
  override class var layerClass: AnyClass { CAGradientLayer.self }

  private var gradient: CAGradientLayer { layer as! CAGradientLayer }

  override init(frame: CGRect) {
    super.init(frame: frame)
    isUserInteractionEnabled = false
  }

  required init?(coder: NSCoder) { nil }

  func update(edge: ProgressiveBlurEdge, color: UIColor, stops: [GradientStop]) {
    gradient.colors = stops.map { color.withAlphaComponent(color.cgColor.alpha * $0.alpha).cgColor }
    gradient.locations = stops.map { NSNumber(value: Double($0.location)) }

    switch edge {
    case .top:
      gradient.startPoint = CGPoint(x: 0.5, y: 1)
      gradient.endPoint = CGPoint(x: 0.5, y: 0)
    case .bottom:
      gradient.startPoint = CGPoint(x: 0.5, y: 0)
      gradient.endPoint = CGPoint(x: 0.5, y: 1)
    case .left:
      gradient.startPoint = CGPoint(x: 1, y: 0.5)
      gradient.endPoint = CGPoint(x: 0, y: 0.5)
    case .right:
      gradient.startPoint = CGPoint(x: 0, y: 0.5)
      gradient.endPoint = CGPoint(x: 1, y: 0.5)
    }
  }
}

File: android/src/main/java/expo/modules/blurview/views/ProgressiveBlurView.kt
package expo.modules.blurview.views

import android.content.Context
import android.content.res.Configuration
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.os.Handler
import android.os.Looper
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.HorizontalScrollView
import android.widget.ScrollView
import com.facebook.react.uimanager.BackgroundStyleApplicator
import com.qmdeve.blurview.util.Utils
import expo.modules.blurview.enums.ProgressiveBlurEdge
import expo.modules.blurview.enums.TintStyle
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import kotlin.math.roundToInt

class ProgressiveBlurView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  private val blurLayer = ProgressiveBlurLayer(context)
  private val fallbackGradient = GradientView(context)
  private val scrollMonitor = ScrollBlurMonitor(resources.displayMetrics.density)
  private val mainHandler = Handler(Looper.getMainLooper())

  private var liveBlurAmount = 1f

  private val scrollViewCheck = object : Runnable {
    override fun run() {
      reattachIfScrollViewLeft()
      mainHandler.postDelayed(this, SCROLL_VIEW_CHECK_INTERVAL_MS)
    }
  }

  var intensity: Float = DEFAULT_INTENSITY
    set(value) {
      field = value
      applyIntensity()
      updateGradients()
      updateLayerVisibility()
    }

  var tint: TintStyle = TintStyle.SYSTEM_ULTRA_THIN_MATERIAL
    set(value) {
      field = value
      updateGradients()
    }

  var blurTintColor: Int? = null
    set(value) {
      field = value
      updateGradients()
    }

  var edge: ProgressiveBlurEdge = ProgressiveBlurEdge.TOP
    set(value) {
      field = value
      blurLayer.edge = value
      updateGradients()
      updateLayerVisibility()
    }

  var startOffset: Float = 0f
    set(value) {
      field = value
      blurLayer.fadeLength = fadeLength
      updateGradients()
    }

  var scrollFallback: Boolean = true
    set(value) {
      field = value
      scrollMonitor.tracksSpeed = value
      updateLayerVisibility()
    }

  var fallbackColor: Int? = null
    set(value) {
      field = value
      updateGradients()
    }

  init {
    addView(blurLayer, 0)
    fallbackGradient.visibility = INVISIBLE
    addView(fallbackGradient, 1)

    scrollMonitor.tracksSpeed = scrollFallback
    scrollMonitor.onScroll = { updateLayerVisibility() }
    scrollMonitor.onContentChange = { updateLayerVisibility() }
    scrollMonitor.onBlurAmountChange = { amount ->
      liveBlurAmount = amount
      updateLayerVisibility()
    }
    applyIntensity()
  }

  private val fadeLength: Float
    get() = 1 - startOffset.coerceIn(0f, 1f)

  private val strength: Float
    get() = intensity.coerceIn(0f, 100f) / 100f

  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    setMeasuredDimension(
      MeasureSpec.getSize(widthMeasureSpec),
      MeasureSpec.getSize(heightMeasureSpec)
    )
  }

  override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
    blurLayer.layoutExactly(0, 0, r - l, b - t)
    fallbackGradient.layoutExactly(0, 0, r - l, b - t)
    updateGradients()

    if (scrollMonitor.scrollView == null) {
      attachScrollMonitor()
    }
    updateLayerVisibility()
  }

  override fun onConfigurationChanged(newConfig: Configuration?) {
    super.onConfigurationChanged(newConfig)
    updateGradients()
  }

  private fun applyIntensity() {
    val radius = strength * BlurView.MAX_RADIUS_DP * resources.displayMetrics.density
    if (radius <= 0f) return
    blurLayer.radius = radius
  }

  private fun updateGradients() {
    val background = backgroundColorBehind()

    val wash = blurTintColor ?: withAlpha(background, DEFAULT_WASH_OPACITY)
    blurLayer.washColor = if (Color.alpha(wash) > 0) wash else tint.toOverlayColor(intensity)

    val fallback = fallbackColor ?: background
    val stops = fade(0f, fadeLength, rising = true) + GradientStop(1f, 1f)
    fallbackGradient.update(edge, fallback, stops.map { GradientStop(it.location, it.alpha * strength) })
  }

  private fun backgroundColorBehind(): Int {
    var ancestor = parent as? View
    while (ancestor != null) {
      val color = runCatching { BackgroundStyleApplicator.getBackgroundColor(ancestor) }.getOrNull()
        ?: (ancestor.background as? ColorDrawable)?.color
      if (color != null && Color.alpha(color) == 255) {
        return color
      }
      ancestor = ancestor.parent as? View
    }
    val typedValue = TypedValue()
    return if (context.theme.resolveAttribute(android.R.attr.colorBackground, typedValue, true) &&
      typedValue.type >= TypedValue.TYPE_FIRST_COLOR_INT &&
      typedValue.type <= TypedValue.TYPE_LAST_COLOR_INT
    ) {
      typedValue.data
    } else {
      Color.WHITE
    }
  }

  private fun updateLayerVisibility() {
    val coverage = contentCoverage()
    val blurAmount = if (scrollFallback) liveBlurAmount else 1f

    val blurAlpha = blurAmount * coverage
    blurLayer.showWith(blurAlpha, blurAlpha > 0f && strength > 0f)

    val fallbackAlpha = (1 - blurAmount) * coverage
    fallbackGradient.showWith(fallbackAlpha, fallbackAlpha > 0f)
  }

  private fun contentCoverage(): Float {
    val scrollView = scrollMonitor.scrollView ?: return 1f
    val content = scrollView.getChildAt(0)
    val contentWidth = (content?.width ?: 0) + scrollView.paddingLeft + scrollView.paddingRight
    val contentHeight = (content?.height ?: 0) + scrollView.paddingTop + scrollView.paddingBottom
    val maxScrollX = (contentWidth - scrollView.width).coerceAtLeast(0)
    val maxScrollY = (contentHeight - scrollView.height).coerceAtLeast(0)

    val hiddenContent = when (edge) {
      ProgressiveBlurEdge.TOP -> scrollView.scrollY
      ProgressiveBlurEdge.BOTTOM -> maxScrollY - scrollView.scrollY
      ProgressiveBlurEdge.LEFT -> scrollView.scrollX
      ProgressiveBlurEdge.RIGHT -> maxScrollX - scrollView.scrollX
    }
    return smoothstep(0f, COVERAGE_FADE_DISTANCE, hiddenContent / resources.displayMetrics.density)
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    attachScrollMonitor()
    mainHandler.removeCallbacks(scrollViewCheck)
    mainHandler.postDelayed(scrollViewCheck, SCROLL_VIEW_CHECK_INTERVAL_MS)
  }

  override fun onDetachedFromWindow() {
    mainHandler.removeCallbacks(scrollViewCheck)
    scrollMonitor.stop()
    super.onDetachedFromWindow()
  }

  private fun attachScrollMonitor() {
    if (!isAttachedToWindow) {
      scrollMonitor.stop()
      return
    }
    scrollMonitor.observe(findScrollViewBehind())
    updateGradients()
    updateLayerVisibility()
  }

  private fun reattachIfScrollViewLeft() {
    val scrollView = scrollMonitor.scrollView ?: return
    if (scrollView.isAttachedToWindow) return
    scrollMonitor.stop()
    attachScrollMonitor()
  }

  private fun findScrollViewBehind(): ViewGroup? {
    val frame = rectInWindow(this)
    var child: View = this
    var container = child.parent as? ViewGroup

    while (container != null) {
      val index = container.indexOfChild(child)
      for (i in index - 1 downTo 0) {
        scrollView(container.getChildAt(i), frame)?.let { return it }
      }
      child = container
      container = container.parent as? ViewGroup
    }
    return null
  }

  private fun scrollView(view: View, frame: Rect): ViewGroup? {
    if (view.visibility != VISIBLE || !Rect.intersects(rectInWindow(view), frame)) return null
    if (view is ScrollView || view is HorizontalScrollView) {
      return view as ViewGroup
    }
    if (view is ViewGroup) {
      for (i in view.childCount - 1 downTo 0) {
        scrollView(view.getChildAt(i), frame)?.let { return it }
      }
    }
    return null
  }

  override fun draw(canvas: Canvas) {
    if (Utils.sIsGlobalCapturing) return
    super.draw(canvas)
  }

  override fun dispatchDraw(canvas: Canvas) {
    if (Utils.sIsGlobalCapturing) return
    super.dispatchDraw(canvas)
  }

  companion object {
    const val DEFAULT_INTENSITY = 50f
    const val RESERVED_CHILDREN = 2

    private const val COVERAGE_FADE_DISTANCE = 24f

    private const val SCROLL_VIEW_CHECK_INTERVAL_MS = 500L

    private const val DEFAULT_WASH_OPACITY = 0.85f

    private fun withAlpha(color: Int, alpha: Float): Int =
      (color and 0x00FFFFFF) or ((Color.alpha(color) * alpha).roundToInt() shl 24)

    private fun rectInWindow(view: View): Rect {
      val location = IntArray(2)
      view.getLocationInWindow(location)
      return Rect(location[0], location[1], location[0] + view.width, location[1] + view.height)
    }

    private fun View.layoutExactly(l: Int, t: Int, r: Int, b: Int) {
      measure(
        MeasureSpec.makeMeasureSpec((r - l).coerceAtLeast(0), MeasureSpec.EXACTLY),
        MeasureSpec.makeMeasureSpec((b - t).coerceAtLeast(0), MeasureSpec.EXACTLY)
      )
      layout(l, t, r, b)
    }

    private fun View.showWith(alpha: Float, visible: Boolean) {
      if (this.alpha != alpha) this.alpha = alpha
      val target = if (visible) VISIBLE else INVISIBLE
      if (visibility != target) visibility = target
    }
  }
}

private class GradientView(context: Context) : View(context) {
  private val paint = Paint()
  private var edge = ProgressiveBlurEdge.TOP
  private var color = Color.TRANSPARENT
  private var stops: List<GradientStop> = emptyList()
  private var shaderDirty = true

  fun update(edge: ProgressiveBlurEdge, color: Int, stops: List<GradientStop>) {
    if (this.edge == edge && this.color == color && this.stops == stops) return
    this.edge = edge
    this.color = color
    this.stops = stops
    shaderDirty = true
    invalidate()
  }

  override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
    super.onSizeChanged(w, h, oldw, oldh)
    shaderDirty = true
  }

  override fun onDraw(canvas: Canvas) {
    if (width == 0 || height == 0 || stops.isEmpty()) return
    if (shaderDirty) {
      paint.shader = createShader()
      shaderDirty = false
    }
    canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
  }

  private fun createShader() = edgeGradient(edge, width, height, color, stops)
}
